azure-keyvault-cryptography/src/main/java/com/microsoft/azure/keyvault/cryptography/EcKey.java

```diff
@@ -0,0 +1,406 @@
+	 * @param curve
+	 * @throws NoSuchAlgorithmException
+	 * @throws InvalidAlgorithmParameterException
+	 */
+	public EcKey(String kid, JsonWebKeyCurveName curve) throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+		this(kid, curve, new BouncyCastleProvider());
+	}
+	
+	/**
+	 * Constructor.
+	 * 
+	 * Generates a new EcKey with the given curve and kid.
+	 * @param kid
+	 * @param curve
+	 * @param provider Java security provider
+	 * @throws InvalidAlgorithmParameterException
+	 * @throws NoSuchAlgorithmException
+	 */
+	public EcKey(String kid, JsonWebKeyCurveName curve, Provider provider) throws InvalidAlgorithmParameterException, NoSuchAlgorithmException {
+		_kid = kid;
+		_provider = provider;
+		_curve = curve;
+		
+		_signatureAlgorithm = CURVE_TO_SIGNATURE.get(curve);
+		if (_signatureAlgorithm == null) {
+			throw new NoSuchAlgorithmException("Curve not supported.");
+		}
+		
+		final KeyPairGenerator generator = KeyPairGenerator.getInstance("EC", provider);
+		ECGenParameterSpec gps = new ECGenParameterSpec(CURVE_TO_SPEC_NAME.get(curve));
+
+		generator.initialize(gps);
+		_keyPair = generator.generateKeyPair();
+		
+	}
+	
+	/**
+	 * Constructor.
+	 * 
+	 * Generates a new EcKey with the given keyPair.
+	 * The keyPair must be an ECKey.
+	 * @param keyPair
+	 * @throws NoSuchAlgorithmException
+	 * @throws InvalidAlgorithmParameterException
+	 */
+	public EcKey(KeyPair keyPair) throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+		this(null, keyPair);
+	}
+	
+	/**
+	 * Constructor.
+	 * 
+	 * Generates a new EcKey with the given keyPair.
+	 * The keyPair must be an ECKey.
+	 * @param kid
+	 * @param keyPair
+	 * @throws NoSuchAlgorithmException
+	 * @throws InvalidAlgorithmParameterException
+	 */
+	public EcKey(String kid, KeyPair keyPair) throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+		this(kid, keyPair, new BouncyCastleProvider());
+	}
+	
+	/**
+	 * Constructor.
+	 * 
+	 * Generates a new EcKey with the given keyPair.
+	 * The keyPair must be an ECKey.
+	 * @param kid
+	 * @param keyPair
+	 * @param provider Java security provider
+	 * @throws NoSuchAlgorithmException
+	 * @throws InvalidAlgorithmParameterException
+	 */
+	public EcKey(String kid, KeyPair keyPair, Provider provider) throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+
+        if (keyPair == null) {
+            throw new IllegalArgumentException("Please provide an ECKey");
+        }
+
+        if (keyPair.getPublic() == null || !(keyPair.getPublic() instanceof ECPublicKey)) {
+            throw new IllegalArgumentException("The keyPair provided is not an ECKey");
+        }
+        
+        _kid      = kid;
+        _keyPair  = keyPair;
+        _provider = provider;
+        _curve = getCurveFromKeyPair(keyPair);
+        _signatureAlgorithm = CURVE_TO_SIGNATURE.get(_curve);
+		if (_signatureAlgorithm == null) {
+			throw new NoSuchAlgorithmException("Curve not supported.");

```

Comment:
>throw new NoSuchAlgorithmException("Curve not supported."); [](start = 3, length = 59)

this exception type seems like it should be a IllegalArgumentException rather than NoSuchAlgorithmException.  The user is specifying the key not the algorithm here.  The message would be the same though.